### PR TITLE
Simplify regex and test on Turkish locale

### PR DESCRIPTION
**Pull request by kant2002**
I remove RegexOptions.IgnoreCase since somehow this break the whole recognition on Turkish locale and I don't see any point have it. We already define both uppercase and lower case letters in the regex. Alternative for A-Za-z is to use \w and for 0-9 use \d, so maybe this will lead to more reliable version recognizer. If strict fragment rules needed they should be placed to RegexOptionalFragment, but I was unable to make this work reliable so far.

See #253

### CODE REVIEW

**Comment by billti** [comment]
Great job figuring this out! The docs do actually call out the ignoreCase can cause issues on Turkish locale (see https://msdn.microsoft.com/en-us/library/yd1hzczs.aspx#Invariant ). Using \w or \d would include non-ASCII characters in a .NET RegEx unless ECMAScript mode is enabled, so using the character ranges explicitly is still the best way to go here I think. Other comments on the code.

File: Nodejs/Product/Npm/SemverVersion.cs [hunk 1]
```diff
@@ -31,13 +31,13 @@ public struct SemverVersion {
             "^(?<major>[0-9]+)"
             + "\\.(?<minor>[0-9]+)"
             + "\\.(?<patch>[…0-9]+)" // The '…' is there to handle the 'classy' library, which has a very long version number - see slightly snarky comment about that unadulterated bag of hilarity below.
-            + "(?:-(?<prerelease>[…0-9A-Za-z-]+(\\.[…0-9A-Za-z-]+)*))?"
-            + "(?:\\+(?<buildmetadata>[…0-9A-Za-z-]+(\\.[…0-9A-Za-z-]+)*))?$",
-            RegexOptions.IgnoreCase | RegexOptions.Singleline);
+            + "(?:-(?<prerelease>[…0-9A-Za-z-.]+))?"
```
Comment (billti) [inline Nodejs/Product/Npm/SemverVersion.cs]: I get adding the "." to allow multi-part pre-release tags (i.e. "1.0.1-beta.1"), but doesn't this also allow invalid formats such as "1.0.1-..5.."? Also, I though the "-" had to be immediately after the opening "[" or before the closing "]" to be treated as a regular character rather than the range operator.

File: Nodejs/Product/Npm/SemverVersion.cs [hunk 2]
```diff
@@ -31,13 +31,13 @@ public struct SemverVersion {
             "^(?<major>[0-9]+)"
             + "\\.(?<minor>[0-9]+)"
             + "\\.(?<patch>[…0-9]+)" // The '…' is there to handle the 'classy' library, which has a very long version number - see slightly snarky comment about that unadulterated bag of hilarity below.
```
Comment (billti) [inline Nodejs/Product/Npm/SemverVersion.cs]: p.s. I still don't get what the '…' actually does here. It's not valid text in a SemVer version or a RegEx special char (that I can find), and I don't see how it helps with the 'classy' library. If someone can expand on how this actually makes the code behave more correctly, I'd appreciate it.

File: Nodejs/Product/Npm/SemverVersion.cs [hunk 3]
```diff
@@ -60,12 +60,26 @@ public struct SemverVersion {
...
+                    preReleaseFragment = preRelease.Value;
+                }
+
+                string buildMetadataFragment;
+                if (!buildMetadata.Success || !IsValidOptionalFragment(buildMetadata.Value)) {
+                    buildMetadataFragment = null;
+                } else {
+                    buildMetadataFragment = buildMetadata.Value;
+                }
+
```
Comment (billti) [inline Nodejs/Product/Npm/SemverVersion.cs]: I don't see what value adding this code gets. The only difference over the original ternary operators is that is also calls IsValidOptionalFragment, but that just checks the fragment matches a RegEx which is identical to the match in the original RegEx that created the "prerelease" capture group. Thus is must be valid, else it wouldn't have been in the capture group and prerelease.Success would be false.

File: Nodejs/Product/Npm/SemverVersion.cs [hunk 4]
```diff
@@ -31,13 +31,13 @@ public struct SemverVersion {
...
-            + "(?:\\+(?<buildmetadata>[…0-9A-Za-z-]+(\\.[…0-9A-Za-z-]+)*))?$",
-            RegexOptions.IgnoreCase | RegexOptions.Singleline);
+            + "(?:-(?<prerelease>[…0-9A-Za-z-.]+))?"
+            + "(?:\\+(?<buildmetadata>[…0-9A-Za-z-.]+))?$",
+            RegexOptions.Singleline);
 
         private static readonly Regex RegexOptionalFragment = new Regex(
-            "^[…0-9A-Za-z-]+(\\.[…0-9A-Za-z-]+)*$",
-            RegexOptions.IgnoreCase | RegexOptions.Singleline);
+            "^[…0-9A-Za-z-.]+$",
```
Comment (billti) [inline Nodejs/Product/Npm/SemverVersion.cs]: Same here and on line 35. The prior pattern of `^[…0-9A-Za-z-]+(\\.[…0-9A-Za-z-]+)*` looks right to me in terms of disallowing invalid formats. `^[…0-9A-Za-z-.]+` seems a little too lax. Why is this change needed?

File: Nodejs/Product/Npm/SemverVersion.cs (hunk 2 above)
Comment (mousetraps) [inline Nodejs/Product/Npm/SemverVersion.cs]: Way back when... npm did not have the db stored in a json format, so we used to parse the textual output ([here's an example](https://raw.githubusercontent.com/Microsoft/nodejstools/1bb20aa7ed8b8ace307825ed7072ea54fb7ff40d/Nodejs/Tests/TestData/NpmSearchData/npmsearchfullcat_npm143.txt))

We've since [removed that code](https://github.com/Microsoft/nodejstools/commit/a1cb641ce7a9bdf7fe240d3ae069bde1f9414a2b), but looks like we missed a spot.

In general, this file needs some attention, but it's one of those areas that's been around for a while, and works well enough.

Comment (mousetraps) [inline Nodejs/Product/Npm/SemverVersion.cs]: (so this was handling the case where the textual output for the version number was cut off by the next column, and therefore ended with a ...)

Comment (billti) [inline Nodejs/Product/Npm/SemverVersion.cs]: Ah, got it. So it's not part of the version number, just part of the textual output the code used to scrape, where it used this char when truncating a column. Makes sense.

It would be good to just delete these if no longer needed. I spent quite a while looking at the SemVer spec, the 'classy' history, and the RegEx operators to trying and figure out what the hell it might be doing :confused:

**Comment by mousetraps** [comment]
Nice job tracking this down! In addition to Bill's comments... let's go ahead and catch this exception during GetCatalog (and print to the output window) so that people aren't left empty handed if parsing fails (as we do with ArgumentException)
https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Product/Npm/SPI/NpmGetCatalogCommand.cs#L143-L148

**Comment by kant2002** [comment]
Probably was too late to debug this issue, now patch is much more clean. I just have to remove `RegexOptions.IgnoreCase`
@mousetraps and @billti please review.

File: Nodejs/Product/Npm/SPI/NpmGetCatalogCommand.cs [hunk 5]
```diff
@@ -120,7 +120,10 @@ internal class NpmGetCatalogCommand : NpmCommand, IPackageCatalog {
                                             .FirstOrDefault();
 
                                         if (!string.IsNullOrEmpty(latestVersion)) {
-                                            builder.LatestVersion = SemverVersion.Parse(latestVersion);
+                                            try {
+                                                builder.LatestVersion = SemverVersion.Parse(latestVersion);
+                                            } catch (SemverVersionFormatException) {
```
Comment (mousetraps) [inline Nodejs/Product/Npm/SPI/NpmGetCatalogCommand.cs]: Also print out a message here - we want to know about the issue, but don't want it to block the user.

**Comment by mousetraps** [comment]
minor comment, but lgtm other than that!

**Comment by mousetraps** [comment]
@kant2002 I presume you mistakenly closed this since it looks like you restored your branch too, so I went ahead ad re-opened it :smile:

**Comment by kant2002** [comment]
True. I just add reporting of the message in case of invalid package semversion
